doc/otp-tokend.1
...\" $Header: /usr/src/docbook-to-man/cmd/RCS/docbook-to-man.sh,v 1.3 1996/06/17 03:36:49 fld Exp $
...\"
...\"	transcript compatibility for postscript use.
...\"
...\"	synopsis:  .P! <file.ps>
...\"
.de P!
\\&.
.fl			\" force out current output buffer
\\!%PB
\\!/showpage{}def
...\" the following is from Ken Flowers -- it prevents dictionary overflows
\\!/tempdict 200 dict def tempdict begin
.fl			\" prolog
.sy cat \\$1\" bring in postscript file
...\" the following line matches the tempdict above
\\!end % tempdict %
\\!PE
\\!.
.sp \\$2u	\" move below the image
..
.de pF
.ie     \\*(f1 .ds f1 \\n(.f
.el .ie \\*(f2 .ds f2 \\n(.f
.el .ie \\*(f3 .ds f3 \\n(.f
.el .ie \\*(f4 .ds f4 \\n(.f
.el .tm ? font overflow
.ft \\$1
..
.de fP
.ie     !\\*(f4 \{\
.	ft \\*(f4
.	ds f4\"
'	br \}
.el .ie !\\*(f3 \{\
.	ft \\*(f3
.	ds f3\"
'	br \}
.el .ie !\\*(f2 \{\
.	ft \\*(f2
.	ds f2\"
'	br \}
.el .ie !\\*(f1 \{\
.	ft \\*(f1
.	ds f1\"
'	br \}
.el .tm ? font underflow
..
.ds f1\"
.ds f2\"
.ds f3\"
.ds f4\"
.ta 8n 16n 24n 32n 40n 48n 56n 64n 72n 
.TH "\fBotp-tokend\fP" "1"
.SH "NAME"
\fBotp-tokend\fP \(em Deliver tokens via SMTP(S) or HTTP(S)\&.
.SH "SYNOPSIS"
.PP
\fBotp-tokend\fP [-?Dv]  [-b\fI bind-path\fP]  [-f\fI from-address\fP]  [-s\fI subject\fP]  [-S\fI smtp_url\fP]  [-P\fI pidfile\fP]  [-H\fI httl_url\fP] 
.SH "DESCRIPTION"
.PP
The \fBotp-tokend\fP daemon processes send-token
requests from the Unix domain socket \fB/var/run/otp-tokend\fP\&.
Tokens are then delivered to the end-user via SMTP, SMTPS, HTTP, or HTTPS\&.
Delivery through SMS or other out of band mechanisms can be implemented
with a HTTPS back end\&.
.PP
Otplib will optionally generate a send-token request when the user record has
the send-token flag and location information set\&.  This can be done with
\fBotp-control\fP\&.
.PP
The send-token request is a single datagram with the following ASCII fields
separated by newlines:
.PP
.nf
service
username
location
token
.fi
.PP
If the location (destination to send the token) is all digits the request is
sent via HTTP(S)\&.  A request with the @ symbol is sent via SMTP(S)\&.  All other
requests are ignored\&.
.SH "OPTIONS"
.IP "-b, --bind-path=\fI bind_path\fP" 10
Listen for send-token requests from an alternate Unix domain socket\&.
.IP "-D, --disable-daemon-mode" 10
Disable daemon mode\&.  When specified \fBotp-tokend\fP will not
run in the background and stdout is available for debugging information\&.
.IP "-f, --from-address=\fI from_address\fP" 10
Specify the e-mail from address when using SMTP(S)\&.
.IP "-s, --subject=\fI subject\fP" 10
Specify the e-mail subject when using SMTP(S)\&.
.IP "-s, --smtp-url=\fI smtp_url\fP" 10
Specify the URL to an SMTP(S) server\&.  Example smtp://127\&.0\&.0\&.1\&.
.IP "-P, --pidfile=\fI pid_file\fP" 10
Specify an alternate location for a file containing the process ID
of the \fBotp-tokend\fP server\&.
.IP "-H, --http-url=\fI http_url\fP" 10
Specify the URL to an HTTP(S) server\&.  A form will be posted to the
url with the form "to" set to the location, and
"msg" set to the token\&.
.IP "-v, --verbose" 10
Increase verbosity\&.  Can be used multiple times\&.
.IP "--version" 10
Display software version\&.
.SH "EXAMPLES"
.PP
Start the \fBotp-tokend\fP server in the foreground with verbose
output using https://127\&.0\&.0\&.1:443/sms for HTTP and and smtp://127\&.0\&.0\&.1 for SMTP requests\&.
.PP
.nf
otp-tokend -D -vvv -S smtp://127\&.0\&.0\&.1 -H https://127\&.0\&.0\&.1:443/sms
.fi
.SH "AUTHOR"
.PP
Mark Fullmer maf@splintered\&.net
.SH "SEE ALSO"
.PP
\fBotp-control\fP(1)
\fBotp-sca\fP(1)
\fBotp-sct\fP(1)
\fBpam_otp\fP(1)
\fBhtsoft-downloader\fP(1)
\fBbcload\fP(1)
\fBotp-ov-plugin\fP(1)
spyrus-par2(7)
...\" created by instant / docbook-to-man, Thu 12 Dec 2013, 10:40


doc/CHANGES
# $Id:$

htsoft-downloader: Makefile install target fix - eshicks

pam_otp: allown_unknown typo - eshicks

otplib: more sanity checking on otp_user_add - eshicks

otp-tokend: smtp support, fork() on message

urd/rc.d/urd: updated for new flags

otp-control: mode generate will honor -c

send-tokendd: reference implementation of OOB token.  Sends token to web
page via ssl.  Used internally for SMS gateway.

send-token feature.  Send to token to user via flexible OOB method.

HOTPC.BAS rev 9.  Check PIN, readerKey

otp-sca support SC_SETREADERKEY_CAP

pam_otp.so: require_db_entry is now defaulted to enabled.  A new option
allow_unknown_user which is mutually exclusive to require_db_entry is
accepted for consistency with urd and openvpn.  kyriacou@osc.edu

otpdb2sc missing quote.  kyriacou@osc.edu

openvpn plugin accepts otp-allow-unknown-user

otp-control will by default create a new user with count=1.  BasicCard
code treats count=0 as a special case when a challenge (count) is supplied.
0 is ignore challenge and use EEPROM stored count.

getopt_long() support.  All binaries support --version.

honor OTP_FLAGS_DISPNT in urd.  Use -c to force display

urd will omit password when packed decode debug is enabled

fix state hex encoding in urd when state > 127

fix soft reload option in urd

add PAM password authentication to urd

str.c: tcsetattr() missing in str_input()

otp-control -m list-sc will automatically set the correct FMT flags

document behavior of count=0 in otp-sct, otp-sca, and spyrus-reader

add urd/rc.d script to dist

add rfc.card to dist

spyrus1.4.hex installed instead of 1.3.  Also in QUICKSTART

### ootp-1.03.tar.gz

add CHANGES and TODO to dist

balance card enabled by default with HOTPC

finish dec31.x and DHEX40 otp-sca, otp-sct

add -i option to htsoft-downloader to ignore last WOK timeout

reorganize BasicCard code to minimize RAM and Flash resources.  Fix
PRDisplay (balance reader) syntax error intruduced with BalanceCardIndex

### snap-12-16-2009

add dhex40 format

add set-flags

remove activate, disable, deactivate, flags-dspcnt, flags-no-dspcnt
from otp-control.  Add more generic set-flags, set-format, set-status,
and set-type.

add dec31.6 dec31.7 dec31.8 dec31.9 dec31.10 formats.

add runtime window option to otp-openvpn, urd, and pam_otp.

scr: default to embedded reader if pcscd is not running.  PCSC: defaults
 to first PCSC reader.

urd: -O not positionally dependent on -o. Can disable OTP and specify otpdb.

add FreeBSD rc.d script for urd

add Zeitcontrol balance reader datasheet to docs

### ootp-1.02.tar.gz

otpdb2sc: do not right 0 pad index to otp-control

pam_otp.so: compile PIC

### ootp-1.01.tar.gz

release: build man and html pages at release time

sccmd.c: SetAdminKey fw version check reversed

urd: build default HOTP enabled

### ootp-1.00.tar.gz


otp-tokend/otp-tokend.c
#include <errno.h>
#include <unistd.h>
#include <stdarg.h>
#include <sys/types.h>
#include <stdio.h>
#include <netdb.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <getopt.h>
#include <stdlib.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <signal.h>
#include <arpa/inet.h>
#include <unistd.h>
#include <string.h>
#include <strings.h>
#include <curl/curl.h>
#include <ctype.h>
#include "xerr.h"
#include "otplib.h"

/* 
 * XXX man page
 */
static void usage(void);
static int write_pidfile(char *fname);

#define REQ_MODE_HTTP 0x1
#define REQ_MODE_SMTP 0x2

#define NXT_FIELD(V1,V2)\
  f = strsep(&c, "\n");\
  if (!f) {\
    xerr_warnx("parse rx_buf fail at %s", V1);\
    continue;\
  }\
  V2 = c;\

size_t curl_write_cb(void *ptr, size_t size, size_t nmemb, void *userdata);
size_t curl_read_cb(void *ptr, size_t size, size_t nmemb, void *userdata);

char *global_token;
char *global_svc;
char *global_hdr_subject;
char *global_hdr_from;

int main(int argc, char **argv)
{
  extern char *ootp_version;
  struct sockaddr_un rx_path;
  pid_t pid_child;
  CURL *curl;
  struct curl_slist *smtp_rcpt = NULL;
  char rx_buf[1024], *c, *f, *msg_svc, *msg_user, *msg_loc, *msg_token;
  char msg_buf[1024], post_buf[1024], *msg_ue, *loc_ue, *rx_pathname;
  char buf[1024], *pid_fname, *url_http, *url_smtp, *url, *hdr_from;
  char *hdr_subject;
  int rx_sock, len, verbose, opt_version, daemon_mode, buf_l, i;
  int req_mode, isdigits, isemail;

  struct option longopts[] = {
    { "bind-path",            1, (void*)0L,    'b'},
    { "disable-daemon-mode",  0, (void*)0L,    'D'},
    { "from-address",         1, (void*)0L,    'f'},
    { "help",                 0, (void*)0L,    'h'},
    { "help",                 0, (void*)0L,    '?'},
    { "subject",              1, (void*)0L,    's'},
    { "smtp-url",             1, (void*)0L,    'S'},
    { "pidfile",              1, (void*)0L,    'P'},
    { "http-url",             1, (void*)0L,    'H'},
    { "verbose",              0, (void*)0L,    'v'},
    { "version",              1, &opt_version,  1},
    { 0, 0, 0, 0},
  };

  req_mode = 0;
  daemon_mode = 1;
  opt_version = 0;
  smtp_rcpt = (struct curl_slist*)0L;
  pid_fname = "/var/run/otp-tokend.pid";
  url_http = (char*)0L;
  url_smtp = (char*)0L;
  url = (char*)0L;
  hdr_from = "hotp@eng.oar.net";
  hdr_subject = "HOTP Token";
  verbose = 0;
  xerr_setid(argv[0]);
  rx_pathname = OTP_SEND_TOKEN_PATHNAME;

  while ((i = getopt_long(argc, argv, "b:Df:h?H:P:s:S:v", longopts,
    (int*)0L)) != -1) {

    switch (i) {

      case 'b':
        rx_pathname = optarg;
        break;

      case 'D':
        daemon_mode = 0;
        break;

      case 'f':
        hdr_from = optarg;
        break;

      case 'h':
      case '?':
        usage();
        exit(0);
        break;

      case 'P':
        pid_fname = optarg;
        break;

      case 's':
        hdr_subject = optarg;
        break;

      case 'S':
        url_smtp = optarg;
        break;

      case 'H':
        url_http = optarg;
        break;

      case 'v':
        ++verbose;
        break;

      case 0:
        if (opt_version) {
          printf("%s\n", ootp_version);
          exit(0);
        }

      default:
        xerr_errx(1, "getopt_long(): fatal.");

    } /* switch */

  } /* while getopt_long() */

  global_hdr_subject = hdr_subject;
  global_hdr_from = hdr_from;

  if (!url_http || !url_smtp)
    xerr_errx(1, "HTTP and SMTP url required.");

  if (daemon_mode) {

    xerr_setsyslog2(1); /* use syslog for output */
      
    /* run in the background */
    if (daemon(0, 0) < 0)
      xerr_err(1, "dameon()");
      
  } /* daemon_mode */

  buf_l = snprintf(buf, sizeof(buf), "tokend start:");
  if (verbose > 1) {
    for (i = 0; i < argc; ++i)
      buf_l += snprintf(buf+buf_l, sizeof(buf)-buf_l, " %s", argv[i]);
  }
  xerr_info(buf);

  bzero(&rx_path, sizeof (rx_path));
  rx_path.sun_family = AF_UNIX;
  if (strlen(rx_pathname) >= sizeof(rx_path.sun_path))
    xerr_errx(1, "rx_pathname too long.");
  strncpy(rx_path.sun_path, rx_pathname, sizeof(rx_path.sun_path));

  /* write out pidfile */
  if (write_pidfile(pid_fname) < 0)
    xerr_errx(1, "write_pidfile(%s): fatal", buf);

  if ((rx_sock = socket(AF_UNIX, SOCK_DGRAM, 0)) < 0)
    xerr_err(1, "socket()");

  /* fail silently */
  umask(077);
  unlink(rx_pathname);

  if (bind(rx_sock, (struct sockaddr*)&rx_path, sizeof(rx_path)) < 0)
    xerr_err(1, "bind(%s)", rx_pathname);

  /* reap children */
  if (signal(SIGCHLD, SIG_IGN) == SIG_ERR)
    xerr_errx(1, "signal(SIGCHLD)");

  while (1) {

    if ((len = recv(rx_sock, &rx_buf, sizeof(rx_buf), 0)) < 0)
      xerr_err(1, "recv()");

    if ((pid_child = fork()) == -1)
      xerr_err(1, "fork()");

    /* parent? */
    if (pid_child)
      continue;

    /* child */
    if (verbose > 2)
      xerr_info("Child pid=%lu.", (unsigned long)getpid());

    if (len == 0)
      xerr_errx(1, "rx_buf empty.");

    if (rx_buf[len - 1] != 0)
      xerr_errx(1, "recv(): rx_buf not null terminated, skipping.");

    c = rx_buf;

    msg_svc = rx_buf;
    NXT_FIELD("msg_user", msg_user);
    NXT_FIELD("msg_loc", msg_loc);
    NXT_FIELD("msg_token", msg_token);

    for (c = msg_token; *c; ++c)
      if (*c == '\n')
        *c = 0;

    /* guess destination.  All digits == http, @ == smtp */
    isdigits = 1;
    isemail = 0;
    for (c = msg_loc; *c; ++c) {
      if (!isdigit(*c))
        isdigits = 0;
      if (*c == '@')
        isemail = 1;
    }
    if (isdigits) {
      req_mode = REQ_MODE_HTTP;
      url = url_http;
    } else if (isemail) {
      req_mode = REQ_MODE_SMTP;
      url = url_smtp;
    } else {
      xerr_errx(1, "Req mode not set for %s.", msg_loc);
    }

    if (curl_global_init(CURL_GLOBAL_ALL) != CURLE_OK)
      xerr_errx(1, "curl_global_init(): failed.");

    if (!(curl = curl_easy_init()))
      xerr_errx(1, "curl_easy_init()");

    if (curl_easy_setopt(curl, CURLOPT_URL, url) != CURLE_OK)
      xerr_errx(1, "curl_easy_setopt(url): failed.");

    if (verbose > 1)
      curl_easy_setopt(curl, CURLOPT_VERBOSE, 1);

    if (req_mode == REQ_MODE_HTTP) {

      if (curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION,
        &curl_write_cb) != CURLE_OK) 
        xerr_errx(1, "curl_easy_setopt(CURLOPT_WRITEFUNCTION): failed.");

      snprintf(msg_buf, sizeof(msg_buf), "%s: %s", msg_svc, msg_token);

      if (!(msg_ue = curl_escape(msg_buf, 0)))
        xerr_errx(1, "curl_escape(%s): failed.", msg_buf);

      if (!(loc_ue = curl_escape(msg_loc, 0))) {
        free(msg_ue);
        xerr_errx(1, "curl_escape(%s): failed.", msg_loc);
      }

      snprintf(post_buf, sizeof(post_buf), "to=%s&msg=%s", loc_ue, msg_ue);

      if (curl_easy_setopt(curl, CURLOPT_POSTFIELDS, post_buf) != CURLE_OK)
        xerr_errx(1, "curl_easy_setopt(CURLOPT_POSTFIELDS, %s): failed.",
          post_buf);

      if (curl_easy_perform(curl) != CURLE_OK)
        xerr_errx(1, "curl_easy_perform(): failed.");

      if (verbose > 1)
        xerr_info("msg_buf=%s", msg_buf);

      curl_easy_cleanup(curl);

      curl_global_cleanup();

    } else if (req_mode == REQ_MODE_SMTP) {

      if (curl_easy_setopt(curl, CURLOPT_MAIL_FROM, hdr_from) != CURLE_OK)
        xerr_errx(1, "curl_easy_setopt(CURLOPT_MAIL_FROM): failed.");

      if (!(smtp_rcpt = curl_slist_append(smtp_rcpt, msg_loc)))
        xerr_errx(1, "curl_slist_append(smtp_rcpt, msg_loc): failed.");

      if (curl_easy_setopt(curl, CURLOPT_MAIL_RCPT, smtp_rcpt) != CURLE_OK)
        xerr_errx(1, "curl_easy_setopt(CURLOPT_MAIL_RCPT): failed.");

      /* needed by read_cb */
      global_token = msg_token;
      global_svc = msg_svc;

      if (curl_easy_setopt(curl, CURLOPT_READFUNCTION,
        &curl_read_cb) != CURLE_OK)
        xerr_errx(1, "curl_easy_setopt(CURLOPT_READFUNCTION): failed.");

      if (curl_easy_perform(curl) != CURLE_OK)
        xerr_errx(1, "curl_easy_perform(): failed.");

      curl_slist_free_all(smtp_rcpt);

    } else {

      xerr_errx(1, "req_mode");

    }
   
    /* exit child */
    if (verbose > 2)
      xerr_info("child exit");
    exit(0);

  } /* forever waiting messages */

} /* main */

size_t curl_write_cb(void *ptr, size_t size, size_t nmemb, void *userdata)
{
  char *c;
  c = malloc((size*nmemb)+1);
  bcopy(ptr, c, size*nmemb);
  c[size*nmemb] = 0;
  xerr_info(c);
  free(c);
  return size*nmemb;
}

size_t curl_read_cb(void *ptr, size_t size, size_t nmemb, void *userdata)
{
  size_t t, r;
  static int cd;

  if (cd == 0) {
    t = size*nmemb;
    r = snprintf(ptr, t,
      "From: %s <HOTP>\r\nSubject: %s\r\n\r\nToken for %s: %s\r\n",
      global_hdr_from, global_hdr_subject, global_svc, global_token);
    ++cd;
    return r;
  } else {
    return 0;
  }

}

void usage(void)
{
  extern char *ootp_version;
         
  fprintf(stderr, "otp-tokend [-?Dhv] [-b bind-path] [-f from-address] [-s subject]\n");
  fprintf(stderr, "           -S smtp-url [-P pidfile] -H http-url\n");

  printf("%s\n", ootp_version);

}

/*
 * function: write_pidfile()
 *
 * Store proces ID in ASII to fname.
 *
 * returns: < 0 : fail
 *            0 : success
 */
int write_pidfile(char *fname)
{
  int fd, buf_l;
  char buf[512];

  buf_l = snprintf(buf, sizeof(buf), "%lu\n", (unsigned long)getpid());

  if ((fd = open(fname, O_WRONLY|O_CREAT|O_TRUNC, 0644)) < 0 ) {
    xerr_warn("open(%s)", fname);
    return -1;
  }

  if (write(fd, buf, buf_l) < 0) {
    xerr_warn("write(%s)", fname);
    close (fd);
    return -1;
  }

  return (close(fd));

} /* write_pidfile */

